### PR TITLE
[RFC] vim-patch:8.0.0702: an error in a timer can make Vim unusable

### PR DESCRIPTION
**Pull request by gmelodie**
**Problem:** An error in a timer can make Vim unusable.
**Solution:** Don't set the error flag or exception from a timer.  Stop a timer if it causes an error 3 out of 3 times. Discard an exception caused inside a timer.
https://github.com/vim/vim/commit/c577d813b7978345dec4310b2d8f5d5624a681f6

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -17142,10 +17144,18 @@ static void f_timer_stopall(typval_T *argvars, typval_T *unused, FunPtr fptr)
...
   timer_T *timer = (timer_T *)data;
+  int save_did_emsg = did_emsg;
+  int save_called_emsg = called_emsg;
+
   if (timer->stopped || timer->paused) {
     return;
   }
 
+  // Don't run timer while dealing with an error
+  if (aborting()) {
```
Comment (bfredl) [inline src/nvim/eval.c]: This is not needed (doesn't correspond to what the `aborting()` check in the vim patch does).

**Comment by LTKills** [comment]
Errors on Travis don't seem to be related to any code in specific. First one is a certificate error while installing neovim on macOS. The other one is a `Couldn't resolve hostname` while downloading `libtermkey`.

```
[52/74] Performing download step for 'libtermkey'
FAILED: cd /home/travis/build/neovim/neovim/deps-downloads/libtermkey && /usr/local/cmake-3.12.4
/bin/cmake -DPREFIX=/home/travis/nvim-deps/build -DDOWNLOAD_DIR=/home/travis/build/neovim
/neovim/deps-downloads/libtermkey -DURL=http://www.leonerd.org.uk/code/libtermkey/libtermkey-
0.21.1.tar.gz -DEXPECTED_SHA256=cecbf737f35d18f433c8d7864f63c0f878af41f8bd0255a3ebb16010dc044d5f 
-DTARGET=libtermkey -DUSE_EXISTING_SRC_DIR=OFF -P /home/travis/build/neovim/neovim/third-
party/cmake/DownloadAndExtractFile.cmake && /usr/local/cmake-3.12.4/bin/cmake -E touch /home/travis
/nvim-deps/build/src/libtermkey-stamp/libtermkey-download

-- file: /home/travis/build/neovim/neovim/deps-downloads/libtermkey/libtermkey-0.21.1.tar.gz
-- downloading...

     src='http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.21.1.tar.gz'
     dst='/home/travis/build/neovim/neovim/deps-downloads/libtermkey/libtermkey-0.21.1.tar.gz'
     timeout='none'

CMake Error at /home/travis/build/neovim/neovim/third-party/cmake/DownloadAndExtractFile.cmake:74 (message):
  error: downloading
  'http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.21.1.tar.gz' failed
    status_code: 6
    status_string: "Couldn't resolve host name"
    log: Curl_ipv4_resolve_r failed for www.leonerd.org.uk


  Couldn't resolve host 'www.leonerd.org.uk'

  Closing connection 0
```

**Comment by justinmk** [comment]
LGTM except for lint stuff.

File: src/nvim/eval.c [hunk 2]
```diff
@@ -17156,8 +17161,23 @@ static void timer_due_cb(TimeWatcher *tw, void *data)
   argv[0].v_type = VAR_NUMBER;
   argv[0].vval.v_number = timer->timer_id;
   typval_T rettv = TV_INITIAL_VALUE;
+  called_emsg = FALSE;
 
   callback_call(&timer->callback, 1, argv, &rettv);
+
+  // Handle error message
+  if (called_emsg && did_emsg) {
```
Comment (justinmk) [inline src/nvim/eval.c]: `did_emsg` is not in the Vim patch but the test fails if I remove it from here.

**Pull request closed without merging**